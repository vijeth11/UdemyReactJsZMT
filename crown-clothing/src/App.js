import {Component} from 'react';
import { useDispatch } from 'react-redux';
import {Route, Routes } from 'react-router-dom';
import Authentication from './routes/Authentication/authentication.component';
import Checkout from './routes/checkout/checkout.component';
import Home from './routes/home/home.component';
import Navigation from './routes/navigation/navigation.component';
import Shop from './routes/shop/shop.component';
import { setCurrentUser } from './store/user/user.action';
import { createUserDocumentFromAuth, onAuthStateChangedListner } from './utils/firebase/firebase.utils';
import { withParams } from './utils/util/withParams.util';

class App extends Component{

  componentDidMount(){
    this.unSubscribe = onAuthStateChangedListner((user) => {
      if(user){
        createUserDocumentFromAuth(user);
      }
      this.props.dispatch(setCurrentUser(user));
    });
  }

  componentWillUnmount(){
    if(this.unSubscribe)this.unSubscribe();
  }

  render() {
    return (
      <Routes>
        <Route path = '/' element={<Navigation/>}>
          <Route index element={<Home/>}/>
          <Route path='shop/*' element={<Shop/>}/>
          <Route path='auth' element={<Authentication/>}/>
          <Route path='checkout' element={<Checkout/>}/>
        </Route>
      </Routes>      
    );
  }
}

/* 
Routes takes a list Route elemnt with the path and element attribute. Path attribute represents the url path 
and element contains the component that needs to be rendered when the path is reached by user in the browser

Route can have nested Route or children Routes which will be rendered when the child path is reached.
for example 
  <Route path='/home' element = {<Home/>}>
    <Route path='shop' element= {<Shop/>}/>
  </Route>
in the above sample to reach to shop we need to go to /home/shop.But even when we reach this path the content of 
Home component will be renderd but not of the Shop. In order to see the content of the Shop we need to add a component
in render part of Home Component called Outlet like bellow
<div>
  <h1> I am Home</h1>
  <Outlet/>
</div>
the conetnt of the Shop will be rendered in place of Outlet same like ReouterOutlet in Angular.

if you want the first child route to be displayed whenever the path of only parent is reached then 
instead of path the attribute in child Route use index like 
<Route index element = {<Shop/>} />
this will render the content of the Shop component in place of Outlet Tag along with content of parent 
componet whenever the parent path is reached. 
note: If you define multiple index children routes only first will be rendered

Context is used to store data at app level so any leaf node (component) can use it by calling context.
Instead of having to pass the data to parent components all the way up to app and then to child component
in sibbling component via props which is too much data passing between component which does not need them.
This is similar to using Store/Service in angular. 

To Pass MultipleContext to a Component we need to wrap the component with extra ContextConsumer and pass the context 
value as an attribute (ex: NavigationComponent) 

we can have nested routes inside of a component but we need to provide a wild path with '*' in main route like Shop
Inside Shop we have multiple child route wrapped with routes component.

Styled Component is a library used to convert styles into a component and use them in the Component 
look at Button and Navigation component for sample
*/
export default withParams(App,()=>({dispatch:useDispatch()}));
