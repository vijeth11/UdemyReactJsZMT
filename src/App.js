import {React,Component} from 'react';
import { CardList } from './component/card-list/card-list.component';
import './App.css';
import { SearchInput } from './component/search-box/search-box.component';

class App extends Component{
  constructor(){
    super();
    this.state={
      searchField:'',
      monsters:[]
    }
  }

  componentDidMount(){
    fetch('https://jsonplaceholder.typicode.com/users')
    .then(response => response.json())
    .then(response => {this.setState({...this.state, monsters:response}); console.log(response);})
  }

  handleChange = (event) => {
    this.setState({...this.state,searchField:event.target.value })
  }

  render(){
    const {monsters, searchField} = this.state;
    const filteredMonsters = monsters.filter(monster => monster.name.toLowerCase().includes(searchField))
    return(    
    <div className="App">
      <h1> Monsters Rolodex </h1>
      <SearchInput
      placeholder="search monsters" 
      handleChange = {this.handleChange}></SearchInput>
      <CardList monsters={filteredMonsters}></CardList>     
    </div>
   ); 
  }
}

//Note: Why cant we update the state directly?

/* the reason why setState needs to be called if we want to update a state bcz
data modification in react works in a unidirection. So when a click event happens 
react listenes to it and checks for what needs to be done. Once it finds out they have called setState 
it executes this method. In this method it uses new value of state passed as argument and 
shallow updates (change the values in state obj only for keys mentioned in new value) the 
state object. After that it calls render function so that react can now use the updated state object and 
re-render dom with new changes. Due to this it is mandatory to update state via setState method and not directly.
We can only set state directly in constructor because it calls render automatically in constructor for first time.
*/

/*
The content of a React component (usually) needs to contain one root element. If we, for example, try to 
define the component App without the outermost div-element it will throw error
*/

/*
Component can return an array of components as well like:
const App = () => {
  return [
    <h1>Greetings</h1>,
    <Hello name="Maya" age={26 + 10} />,
    <Footer />
  ]
}
instead of returning html
*/

/*
 react needs keyword 'key'  when dealing with list is because. It gives react a knowledge to re-render only the part 
 of the html with the 'key' for which data has been changed in the state by a function or third party due to which the
 heavy lifting of re-rendering each and every elemnt in the list just for change of one elemnt in an array is escaped
 thus making react smoother. 
*/

/*
npm run eject command creates a prod build version of the react application where everything is compiled into js and html 
files. (like ng b --prod in angular)
*/

/*
componentDidMount is the function which is called when the component is mounted for the first time and it just like ngOnInit
in angular. so it runs only once at during component mounting
*/

/*
  order of execution
  Constructor->render->componentDidMount
*/

/*
if a console.log method is added to display state right after the setState method we cannot see the change in the state object
as the setState function is async function we need to pass a callback function as second argument to setState and retrieve new updated state
example setState({...state,searchField:name},() => console.log(this.state))
*/

/*
Arrow function will lexically scope the source class instance to 'this' keyword meaning instance of the class where the arrow function
is defined. on other hand the normal function inside JS class when passed as an argument it does not bing the source class instance to 'this'
keyword used in the function due to where it will throw error when called so in Order to over come this we can add .bind(this) to the function 
while passing as argument so that the source class instance is set as the context of 'this' keyword when called.   
*/
export default App;
