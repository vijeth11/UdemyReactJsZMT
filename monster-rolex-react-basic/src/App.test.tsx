import { shallow } from 'enzyme';
import App from './App';

it("initial test",() => {
    expect(shallow(<App/>)).toMatchSnapshot();
})

it('correctly filters the cards',() => {
    const wrapper = shallow(<App/>);
    
})