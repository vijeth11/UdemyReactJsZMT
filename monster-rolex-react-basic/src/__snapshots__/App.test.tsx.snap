// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial test 1`] = `
<div
  className="App"
>
  <h1>
     Monsters Rolodex 
  </h1>
  <SearchInput
    handleChange={[Function]}
    placeholder="search monsters"
  />
  <CardList
    monsters={Array []}
  />
</div>
`;
